`redundant_allocation` should not be marked as machine applicable

Same reasoning as https://github.com/rust-lang/rust-clippy/issues/6241; fixing this requires non-local changes which can't be done with span suggestions.

I tried this code (https://github.com/rust-lang/rust/blob/0da6d42f297642a60f2640ec313b879b376b9ad8/compiler/rustc_interface/src/queries.rs#L109-L111):

```rust
    fn codegen_backend(&self) -> &Lrc<Box<dyn CodegenBackend>> {
        &self.compiler.codegen_backend()
    }
```

I expected to see this happen: Clippy warns, but doesn't try to fix the suggestion.

Instead, this happened: Clippy tries to fix it, causing an error.

```
error[E0614]: type `(dyn rustc_codegen_ssa::traits::CodegenBackend + 'static)` cannot be dereferenced
   --> compiler/rustc_interface/src/queries.rs:284:43
    |
284 |                 Ok(passes::start_codegen(&***self.codegen_backend(), tcx, &*outputs.peek()))
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: usage of `Rc<Box<T>>`
  --> compiler/rustc_interface/src/interface.rs:47:39
   |
47 |     pub fn codegen_backend(&self) -> &Lrc<Box<dyn CodegenBackend>> {
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Rc<dyn CodegenBackend>`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_allocation
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (ffa2e7a 2020-10-24)

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"royrustdev"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/clippy_lints/src/types/redundant_allocation.rs
+++ b/clippy_lints/src/types/redundant_allocation.rs
@@ -10,6 +10,7 @@ use rustc_span::symbol::sym;
 use super::{utils, REDUNDANT_ALLOCATION};
 
 pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_>, def_id: DefId) -> bool {
+    let mut applicability = Applicability::MaybeIncorrect;
     let outer_sym = if Some(def_id) == cx.tcx.lang_items().owned_box() {
         "Box"
     } else if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
@@ -21,7 +22,6 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
     };
 
     if let Some(span) = utils::match_borrows_parameter(cx, qpath) {
-        let mut applicability = Applicability::MaybeIncorrect;
         let generic_snippet = snippet_with_applicability(cx, span, "..", &mut applicability);
         span_lint_and_then(
             cx,
@@ -63,7 +63,6 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
         None => return false,
     };
     if inner_sym == outer_sym {
-        let mut applicability = Applicability::MaybeIncorrect;
         let generic_snippet = snippet_with_applicability(cx, inner_span, "..", &mut applicability);
         span_lint_and_then(
             cx,

